### DISCUSSION TITLE
useFieldArray - How to persist values changes when conditionally rendering form fields

### DISCUSSION THREAD

**Discussion by knoefel** [react-hook-form/react-hook-form#4565] [Q&A]
Hi! ✌️

I am using `useFieldArray` and would like to render the associated `fields` in different components depending on the viewport (mobile, desktop). Unfortunately, changes in the form fields are not saved when I change the layout (increase/decrease the viewport width) and thus components are re-rendered.  I had assumed that if I set `shouldUnregister` to `false`, then the modified values would also be saved when I then use new form fields with the same name. Is this a mistake on my part? If so, is there any way to get this behaviour?

Steps to reproduce:

1. open [CodeSandbox](https://codesandbox.io/s/usefieldarray-shouldunregister-od64j)
2. modify one of the form fields
3. change viewport width to trigger re-render
4. modified field has initial value again

As always, thanks in advance for clarification and support!!!

**Comment by bluebill1049** [comment]
`shouldUnregister:false`, will keep value even when inputs are unmounted. This will be the default behavior for v7.

**Reply by knoefel** [reply]
Ok. But the CodeSandbox Demo uses `shouldUnregister:false`, that’s why I thought it would keep the modified input value. So my question is: can I somehow persist the modification of the input values or should I use some kind of state management to achieve this?

**Reply by bluebill1049** [reply]
modification of the input, you mean capture the previous and etc?

**Reply by knoefel** [reply]
Yes. For example, if I change the first input value, than alter the viewport width, so that the other inputs (but with the same `name`) are rendered. Than the modified value should persist, but right now I get the initial value again.

**Comment by bluebill1049** [comment]
v7 should resolve your probblem: https://codesandbox.io/s/usefieldarray-shouldunregister-forked-cgw9w?file=/src/App.tsx

**Reply by knoefel** [reply]
Ok, cool. Thanks. 

I noticed something else when updating to V7, maybe a bug. I have created a new discussion -> #4574.

**Comment by knoefel** [comment]
Update: sorry to bother you again, but it isn't working when i use `Controller` to render Material UI TextField components (which is a requirement in my case). I get the same initial result, that the modified value will reset if re-rendering occurs. I updated the CodeSandbox, here is the [Link](https://codesandbox.io/s/usefieldarray-shouldunregister-with-mui-textfield-5k7lk).

**Reply by bluebill1049** [reply]
what's the steps to reproduce?

**Reply by knoefel** [reply]
The same as the first problem:

1. open updated [CodeSandbox](https://codesandbox.io/s/usefieldarray-shouldunregister-with-mui-textfield-5k7lk)
2. modify one of the form fields
3. change viewport width to trigger re-render
4. modified field has initial value again

**Reply by bluebill1049** [reply]
the problem is feilds only contain the defaultValue, you may want to merge with getValues() which is current form values.

**Reply by knoefel** [reply]
Ok, thx. I got it working after adjusting with your last comment. I just replaced `defaultValue={name}` with ```defaultValue={getValues(`data.${index}.name` as const)}```. 
I think you meant something like this when you suggested to merge with `getValues()`?

There is one other strange thing i noticed while updating to v7. On the first re-render the fields array is empty, both on calls with `watch()` and `getValues()`. You can see it in the updated [CodeSandbox](https://codesandbox.io/s/usefieldarray-shouldunregister-with-mui-textfieldedit-tefbn?file=/src/App.tsx:875-930). Just load the CodeSandbox and check the console to see that on the second render it outputs the following:

<img width="1589" alt="Screenshot 2021-04-02 at 12 00 30" src="https://user-images.githubusercontent.com/2396344/113406102-34d50b00-93ab-11eb-8a0f-8218f87e84cc.png">

This wasn't the case with v6 - has there something changed that i'm not aware of? I couldn't find any information regarding this in the migration guide for v7.

**Reply by bluebill1049** [reply]
looks like a bug to me, i am investigating the root cause.

**Reply by bluebill1049** [reply]
this is due to your conditional re-render, pull those out from your wrapper component and it's working correctly now.
<img width="1417" alt="Screen Shot 2021-04-02 at 9 25 48 pm" src="https://user-images.githubusercontent.com/10513364/113407887-ffa3d980-93f9-11eb-89a3-0b23d5809ac4.png">

**Reply by knoefel** [reply]
So you are saying that a conditional render is a problem? Because i checked, that this also isn't working without using a conditional render. If i don't render the `fields`, `watch()` and `getValues()` will be empty on second render. This only happens in v7, with v6 i don't had any problems with that.

[CodeSandbox using v6](https://codesandbox.io/s/usefieldarray--empty-arrayv6-s0n9p)
[CodeSandbox using v7](https://codesandbox.io/s/usefieldarray--empty-arrayv7-syv7j)

Please check the console statements in both CodeSandboxes.

Just for clarification and to better understand my problem here is a short description how i use RHF in my application:

I have a wizard form with 3 steps. It will only render the current step (which the user has selected). One of these steps has a Field Array. Initially it isn't rendered the step with the Field Array. So since v7 thats a problem, because the Field Array will be empty after the first render of the form.

**Reply by bluebill1049** [reply]
here is the problem:

first render: return defaultValue
second render: nothing registered in your use case, hence watch() return empty

works in v6, because v6 have no concept of mounted or not.

**Reply by knoefel** [reply]
Ok, but is there any way to accomplish my use case then? Or can i just not use RHF v7 with conditional rendered parts of my form?

**Reply by bluebill1049** [reply]
I think you will need to let your form rendered once, or why not only sort out the form with css property than completely re-render the form.

**Reply by bluebill1049** [reply]
Let me explain why as well:

- first render watch will return defaultValues or defaultValue from useWatch/watch
- then your component is getting to the render phase
- here we will need to register to invoked and getting all the inputs during the render
- once it's render is ended all inputs reference are inside hook form **(This is the step missing)**
- watch will return what's in the view (which registered)

**Reply by bluebill1049** [reply]
perhaps always render the smallest viewport? so it's mobile first.

**Reply by knoefel** [reply]
Thanks for the explanation! 

Unfortunately, hiding the steps which are not currently displayed is not an option in my case. The whole form is quite complex with many fields which are displayed inside a dialog:

1st step: 20 fields
2nd step: 15 field arrays with each 2 fields, often with many entries
3rd step: field array with each 10 fields, there can also be 20+ entries

This means that the total number of fields can quickly reach 400+. In an earlier version of the app i was doing it like you suggested. But with all the field arrays, each with many fields of its own, and them additionally also packed in Material UI Accordions this caused mobile performance problems, especially with the amount of all the fields only hidden via CSS.

Probably another solution would be to manage the form state externally (split the 3 steps up to use `useForm` for each step separately) and no longer use RHF for form fields state management (similar to the Form Wizard in the "Advanced" section of the documentation). Too bad, I really liked RHF's approach in v6 that the form state is managed internally (for unmounted fields). 

Just out of curiosity, what was the reason to move away from the v6 implementation? After all, there was the `shouldUnregister` option for that in v6, which could just be set to `false`. And as I understood the discussion about v7 correctly, this is the default behavior of v7? And i thought this would cover my use case.

I think I have to stay with v6 unfortunately due to my use case, but look forward to working with the new version in a next project. Anyway, Congratulations on the new version!!! 🥳 

And as always thanks for the helpful and super fast support!!!!

**Reply by bluebill1049** [reply]
Thanks @knoefel 🙏 

> Just out of curiosity, what was the reason to move away from the v6 implementation? After all, there was the shouldUnregister option for that in v6, which could just be set to false. And as I understood the discussion about v7 correctly, this is the default behavior of v7? And i thought this would cover my use case.

one of the most important reasons is by supporting both options adds a lot more complexity to the codebase and more error-prone. Also with `shouldUnregister: false`, we will have to `deepClone` `defaultValues`, it gets slower over time once data gets fairly large and deeply nested. so I try to avoid deep clone at all in v7 and keep it simple, light and fast, and obviously, it's not a perfect solution to solve everything, just like what you have run in above. 

also let me share another insight as well with `watch`, why i introduced `mounted` in the codebase, because defaultValue should only return in the first render, just like input's defaultValue, it gets ugly when you remove item and defaultValue returned instead of undefined, it's working in v6, but we had to add more checking and code to support that which i was really hoping to remove them in v7.

**Reply by knoefel** [reply]
Thank you very much for the insights!!!

Unfortunately I have to bother you again, sorry. 👻 You had suggested the solution to my initial problem of this discussion in this [comment](https://github.com/react-hook-form/react-hook-form/discussions/4565#discussioncomment-561635) as follows: 

> the problem is fields only contain the defaultValue, you may want to merge with getValues() which is current form values.

I had then defined the defaultValue as follows to fix the problem:

```defaultValue={getValues(`data.${index}.name` as const)}```

But unfortunately there is maybe a bug. When I want to remove entries of the field array via "remove" method the wrong item gets removed.

Steps to reproduce

1. open [CodeSandbox](https://codesandbox.io/s/usefieldarray-remove-bug-oeu67?file=/src/App.tsx:935-990)
2. remove the first item
3. first item is still there, instead the last item is removed

This happens in v6 as well as in v7.

**Reply by bluebill1049** [reply]
Thanks @knoefel i will take a look at it.

**Reply by bluebill1049** [reply]
follow it here: https://github.com/react-hook-form/react-hook-form/issues/4654

**Reply by bluebill1049** [reply]
can you share some stesps? https://codesandbox.io/s/usefieldarray-remove-bug-forked-7n420

**Reply by knoefel** [reply]
I'm so so sorry, there is no issue. Very embarrassing, but I had actually used the `index` instead of the field `id`. Wow. And that even though I had read the documentation. 🙈😳 

So thanks for your help as always and again sorry for wasting your time.

**Reply by bluebill1049** [reply]
haha no worries.

**Reply by Noitidart** [reply]
> . Also with `shouldUnregister: false`, we will have to `deepClone` `defaultValues`, it gets slower over time once data gets fairly large and deeply nested.

Just curious, would something like `immer` help here? Where we can keep a base and then keep modifying just the affected things? This prevents the need for a merge?

**Comment by Noitidart** [comment]
I think I had the same issue - https://github.com/react-hook-form/react-hook-form/discussions/5100 - maybe I should close that as duplicate? It has a super simplified reproducible demo.

**Reply by bluebill1049** [reply]
left a note in there.
